src/kani-compiler/rustc_codegen_kani/src/overrides/hooks.rs

```diff
@@ -146,13 +146,51 @@ impl<'tcx> GotocHook<'tcx> for Assume {
     }
 }
 
+struct Assert;
+impl<'tcx> GotocHook<'tcx> for Assert {
+    fn hook_applies(&self, tcx: TyCtxt<'tcx>, instance: Instance<'tcx>) -> bool {
+        matches_function(tcx, instance, "KaniAssert")
+    }
+
+    fn handle(
+        &self,
+        tcx: &mut GotocCtx<'tcx>,
+        _instance: Instance<'tcx>,
+        mut fargs: Vec<Expr>,
+        _assign_to: Option<Place<'tcx>>,
+        target: Option<BasicBlock>,
+        span: Option<Span>,
+    ) -> Stmt {
+        assert_eq!(fargs.len(), 2);
+        let cond = fargs.remove(0).cast_to(Type::bool());
+        let msg = fargs.remove(0);
+        let msg = utils::extract_const_message(&msg).unwrap();
+        let target = target.unwrap();
+        let loc = tcx.codegen_span_option(span);
+        let caller_loc = tcx.codegen_caller_span(&span);
+
+        // Since `cond` might have side effects, assign it to a temporary
+        // variable so that it's evaluated once, then assert and assume it
+        let tmp = tcx.gen_temp_variable(cond.typ().clone(), loc.clone()).to_expr();
+        Stmt::block(
+            vec![
+                Stmt::decl(tmp.clone(), Some(cond), loc.clone()),
+                Stmt::assert(tmp.clone(), &msg, caller_loc.clone()),
+                Stmt::assume(tmp, loc.clone()),
```

Comment:
I'm assuming the assumption here is to abort when the assertion fails, right?

Reply:
Yes, this is to maintain the current semantics of assert, which are currently roughly implemented as:
```rust
if !cond {
    assert(false);
    abort();
}
// cond must be true here

Reply:
Can you use the original method `codegen_fatal_error()` instead?

Reply:
As far as I understand, `codegen_fatal_error` is meant for a `panic` statement, which unconditionally fails, whereas here, there is a condition involved.

The alternative I was considering is to do something similar to how rustc expands the `assert` macro, e.g. 
```rust
cond.not().if_then_else(codegen_fatal_error(...), None)
```
However, based on @danielsn's feedback, with an "if", we lose the structure of the original assert.

Reply:
Gotcha. I was just hoping that we could encapsulate and formalized what an assert in the rust / kani world means, which is different from the goto-c semantics. But we can do that in a follow up PR.

Maybe we can do that when we fix codegen `TerminatorKind::Assert()`.